teos-common/src/lightning/ser_macros.rs

```diff
@@ -0,0 +1,390 @@
+                match Readable::read(&mut tracking_reader) {
+                    Err(DecodeError::ShortRead) => {
+                        if !tracking_reader.have_read {
+                            break;
+                        } else {
+                            return Err(DecodeError::ShortRead);
+                        }
+                    },
+                    Err(e) => return Err(e),
+                    Ok(t) => t,
+                }
+            };
+
+            // Types must be unique and monotonically increasing:
+            match last_seen_type {
+                Some(t) if typ.0 <= t => {
+                    return Err(DecodeError::InvalidValue);
+                },
+                _ => {},
+            }
+            last_seen_type = Some(typ.0);
+
+            // Finally, read the length and value itself:
+            let length: BigSize = Readable::read(&mut stream_ref)?;
+            let mut s = ser_utils::FixedLengthReader::new(&mut stream_ref, length.0);
+            match typ.0 {
+                $($type => {
+                    ser_macros::decode_tlv!(s, $field, $fieldty);
+                    if s.bytes_remain() {
+                        s.eat_remaining()?; // Return ShortRead if there's actually not enough bytes
+                        return Err(DecodeError::InvalidValue);
+                    }
+                },)*
+                x if x % 2 == 0 => {
+                    return Err(DecodeError::UnknownRequiredFeature);
+                },
+                _ => {},
+            }
+            s.eat_remaining()?;
+        }
+    } }
+}
+
+macro_rules! init_tlv_field_var {
+    ($field: ident, opt) => {
+        let mut $field = None;
+    };
+    ($field: ident, vec) => {
+        let mut $field = Vec::new();
+    };
+    ($field: ident, opt_str) => {
+        let mut $field = None;
+    };
+}
+
+macro_rules! impl_writeable_msg {
+    ($st: ty, {$($field: ident),* $(,)*}, {$(($type: expr, $tlvfield: ident, $fieldty: tt)),* $(,)*}) => {
+        impl lightning::util::ser::Writeable for $st {
+            fn write<W: lightning::util::ser::Writer>(&self, w: &mut W) -> Result<(), lightning::io::Error> {
+                $(self.$field.write(w)?;)*
+                $crate::lightning::ser_macros::encode_tlv_stream!(w, {$(($type, self.$tlvfield, $fieldty)),*});
+                Ok(())
+            }
+        }
+
+        impl lightning::util::ser::Readable for $st {
+            fn read<R: lightning::io::Read>(r: &mut R) -> Result<Self, lightning::ln::msgs::DecodeError> {
+                $(let $field = lightning::util::ser::Readable::read(r)?;)*
+                $($crate::lightning::ser_macros::init_tlv_field_var!($tlvfield, $fieldty);)*
+                $crate::lightning::ser_macros::decode_tlv_stream!(r, {$(($type, $tlvfield, $fieldty)),*});
+                Ok(Self {
+                    $($field),*,
+                    $($tlvfield),*
+                })
+            }
+        }
+
+        #[cfg(test)]
```

Comment:
Any reason why this is test only? Also, why not derive this for the messages? Is it so you can test it generally within `ser_macros.rs`?

Reply:
I didn't need to impl it at the beginning for usual functionality, and saw `LDK` also doesn't implement it for their messages. But then I needed it for tests, that's why I made it test only.

`ser_macros` wouldn't use it because it tests primitive types which already implement Eq. Generally, tests in `ser_macros` don't test `impl_writeable_msg` but test the macros used inside of it.